tools/codegen-diff-revisions.py

```diff
@@ -0,0 +1,232 @@
+            diff_cmd = f"diff2html -s line -f html -d word -i command --hwt "\
+                f"{tmp_file.name} -F {OUTPUT_PATH}/{dest_path} -- "\
+                f"-U20 {whitespace_flag} {BASE_BRANCH_NAME} {HEAD_BRANCH_NAME} -- {path_to_diff}"
+            eprint(f"Running diff cmd: {diff_cmd}")
+            run(diff_cmd)
+        return dest_path
+
+
+def diff_link(diff_text, empty_diff_text, diff_location, alternate_text, alternate_location):
+    if diff_location is None:
+        return empty_diff_text
+    else:
+        return f"[{diff_text}]({CDN_URL}/{diff_location}) ([{alternate_text}]({CDN_URL}/{alternate_location}))"
+
+
+def make_diffs(base_commit_sha, head_commit_sha):
+    sdk_ws = make_diff("AWS SDK", f"{OUTPUT_PATH}/aws-sdk", base_commit_sha,
+                       head_commit_sha, "aws-sdk", whitespace=True)
+    sdk_nows = make_diff("AWS SDK", f"{OUTPUT_PATH}/aws-sdk", base_commit_sha, head_commit_sha,
+                         "aws-sdk-ignore-whitespace", whitespace=False)
+    server_ws = make_diff("Server Test", f"{OUTPUT_PATH}/codegen-server-test", base_commit_sha,
+                          head_commit_sha, "server-test", whitespace=True)
+    server_nows = make_diff("Server Test", f"{OUTPUT_PATH}/codegen-server-test", base_commit_sha,
+                            head_commit_sha, "server-test-ignore-whitespace", whitespace=False)
+
+    sdk_links = diff_link('AWS SDK', 'No codegen difference in the AWS SDK',
+                          sdk_ws, 'ignoring whitespace', sdk_nows)
+    server_links = diff_link('Server Test', 'No codegen difference in the Server Test',
+                             server_ws, 'ignoring whitespace', server_nows)
+    return "A new generated diff is ready to view.\n"\
+        f"- {sdk_links}\n"\
+        f"- {server_links}\n"
+
+
+def write_to_file(path, text):
+    with open(path, "w") as file:
+        file.write(text)
+
+
+# Prints to stderr
+def eprint(*args, **kwargs):
+    print(*args, file=sys.stderr, **kwargs)
+
+
+# Checks a command result for success and exits the process if it failed
+def expect_cmd_success(command, result):
+    if result.returncode != 0:
+        eprint(f"Command `{command}` failed with status code {result.returncode}")
+        eprint("stderr:\n" + result.stderr.decode("utf-8"))
+        eprint("stdout:\n" + result.stdout.decode("utf-8"))
+        sys.exit(1)
+
+
+# Runs a shell command
+def run(command):
+    result = subprocess.run(command, stdout=sys.stderr, stderr=sys.stderr, shell=True)
+    expect_cmd_success(command, result)
+
+
+# Returns the output from a shell command. Bails if the command failed
+def get_cmd_output(command):
+    result = subprocess.run(command, capture_output=True, shell=True)
+    expect_cmd_success(command, result)
+    return result.stdout.decode("utf-8").strip()
+
+
+# Runs a shell command and returns its exit status
+def get_cmd_status(command):
+    return subprocess.run(command, capture_output=True, shell=True).returncode
```

Comment:
`shell=True` has some security implications: https://docs.python.org/3/library/subprocess.html#security-considerations

Although the usages currently in this script are fine since it's always a static string, I might use `shlex.split` for all of the examples that _don't_ need pipes and only set `shell=True` for the one command that needs pipes.